internal/ingress/annotations/annotations_test.go

```diff
@@ -55,8 +55,14 @@ func TestIngressClassValidatorFunc(t *testing.T) {
 	ing.SetAnnotations(data)
 	for _, test := range tests {
 		ing.Annotations[ingressClassKey] = test.ingress
-		f := IngressClassValidatorFunc(test.controller, test.skipClasslessIngress)
-		b := f(&ing.ObjectMeta)
+		f := IngressClassValidatorFunc(test.controller)
```

Comment:
I suggest rewriting this test to have the following values in the table:
- annotation value
- kong's ingress class specified in config
- selected matching strategy
- expected verdict

This would mean directly putting `ClassHandling` in the table, as opposed to `skipClasslessIngress` (which adds apparently unnecessary complexity and makes it easier to miss missing cases or mistakes)

Reply:
That is essentially what it has now--I've added comments to clarify what's what.

What we don't have is tests of the strategies independent of their user-facing controls: the latter are boolean flags, and the former are derived from them. Effectively we have integration tests here, and a poor delineation between integration and unit tests throughout the codebase, alongside the poor separation between components (the store/annotation distinction is messy in both the tests and implementation).

Reply:
> That is essentially what it has now

My point is that this logic is unnecessary:
```go
	var b bool
		if test.skipClasslessIngress {
			b = f(&ing.ObjectMeta, ExactClassMatch)
		} else {
			b = f(&ing.ObjectMeta, ExactOrEmptyClassMatch)
		}
```

it could be as simple as
```go
    b = f(&ing.ObjectMeta, test.matchingStrategy)
```

> Effectively we have integration tests here

The problem here is that we have replicated a part of the "system under test" (namely, the `if` as quoted above) in the test itself. This is not right because it adds a degree of freedom to the actual system under test. Therefore, IMO it's better to remove the `if` and update the test data to use a `ClassMatching`, not a `bool` as input.